File: rag/app/rag/validator.py
import re
import logging
from typing import Dict, Any

logger = logging.getLogger(__name__)

class VerseValidationError(Exception):
    """Custom exception for Bible archival integrity failures."""
    pass


def is_retrievable_bible_record(metadata: Dict[str, Any]) -> bool:
    """
    Minimal retrieval-time validator.
    Only rejects obvious garbage so the semantic retrieval can still see candidates.
    """
    ga = (metadata.get("ga") or "").strip()
    en = (metadata.get("en") or "").strip()

    bad_markers = [
        "Bible in Ga Language",
        "© Bible Society of Ghana",
        "Currently Selected",
        "Learn More",
        "Rights in the",
    ]

    if any(marker.lower() in ga.lower() for marker in bad_markers):
        return False
    if any(marker.lower() in en.lower() for marker in bad_markers):
        return False

    # allow records that have at least one real side present
    return bool(ga or en)


def is_formattable_bible_record(metadata: Dict[str, Any]) -> bool:
    """
    Strict quote-time validator.
    Only passes records that have all necessary fields to format a perfect quote block.
    """
    required = [
        "reference_display",
        "ga_version_name",
        "ga_version_abbr",
        "english_version_name",
        "english_version_abbr",
        "ga",
        "en",
        "source_name",
    ]
    return all(bool((metadata.get(k) or "").strip()) for k in required)


def format_bible_quote(metadata: Dict[str, Any]) -> str:
    """
    Deterministically formats a validated Bible quote.
    Bypasses the LLM entirely to guarantee 100% template fidelity.
    """
    from app.rag.indexer import BibleRefiner
    
    # DYNAMIC RECONSTRUCTION: If reference_display is missing, build it using BibleRefiner logic
    ref = metadata.get("reference_display")
    if not ref:
        book = metadata.get("book", "Genesis")
        trad_book = metadata.get("traditional_book") or BibleRefiner.get_trad_book(book)
        
        ch_num = int(metadata.get("chapter_num", 0))
        ga_ch = metadata.get("chapter_title_ga") or BibleRefiner.get_ga_chapter_title(ch_num)
        
        v_num = int(metadata.get("verse_num", 0))
        ga_v = metadata.get("ga_label") or BibleRefiner.get_ga_label(v_num)
        
        v_ref = metadata.get("verse_ref") or f"{book} {ch_num}:{v_num}"
        ref = f"{trad_book}, {ga_ch}, {ga_v} ({v_ref})"
    
    ga_name = metadata.get("ga_version_name") or "Ŋmalɛ Krɔŋkrɔŋ Lɛ"
    ga_abbr = metadata.get("ga_version_abbr") or "NEGAB"
    ga_text = metadata.get("ga", "").strip()
    
    en_name = metadata.get("english_version_name") or "King James Version"
    en_abbr = metadata.get("english_version_abbr") or "KJV"
    en_text = metadata.get("en", "").strip()
    
    # Ensure source_name is the filename and strip extension for archival cleanliness
    source = metadata.get("source_name") or metadata.get("filename") or "Bible Archive"
    if "." in source:
        source = ".".join(source.split(".")[:-1])
    source = source.replace("_", " ").strip()

    # Strict multi-line format with identical spacing/labels
    return (
        f"{ref}\n"
        f"Ga Version: {ga_name} ({ga_abbr}):\n"
        f"{ga_text}\n\n"
        f"English Version: {en_name} ({en_abbr}):\n"
        f"{en_text}\n\n"
        f"Source: {source}"
    )

File: rag-cloudflare/src/rag/validator.ts
/**
 * Bible record validators + deterministic quote formatter (ported from app/rag/validator.py).
 */
import type { VerseRecord } from "../types";
import { getGaChapterTitle, getGaLabel } from "./ga";

const BAD_MARKERS = [
  "bible in ga language",
  "© bible society of ghana",
  "currently selected",
  "learn more",
  "rights in the",
];

function str(v: unknown): string {
  return (v == null ? "" : String(v)).trim();
}

/** Minimal retrieval-time validator: reject obvious garbage only. */
export function isRetrievableBibleRecord(meta: Record<string, unknown>): boolean {
  const ga = str(meta.ga).toLowerCase();
  const en = str(meta.en).toLowerCase();
  if (BAD_MARKERS.some((m) => ga.includes(m))) return false;
  if (BAD_MARKERS.some((m) => en.includes(m))) return false;
  return !!(str(meta.ga) || str(meta.en));
}

const REQUIRED = [
  "reference_display",
  "ga_version_name",
  "ga_version_abbr",
  "english_version_name",
  "english_version_abbr",
  "ga",
  "en",
  "source_name",
];

/** Strict quote-time validator: all fields needed for a perfect quote block. */
export function isFormattableBibleRecord(meta: Record<string, unknown>): boolean {
  return REQUIRED.every((k) => str(meta[k]).length > 0);
}

const TRAD_BOOK_FALLBACK: Record<string, string> = {
  genesis: "Mose klɛŋklɛŋ wolo",
  exodus: "Mose wolo ni ji enyɔ",
  leviticus: "Mose wolo ni ji etɛ",
  numbers: "Mose wolo ni ji ejwɛ",
  deuteronomy: "Mose wolo ni ji enumɔ",
};

/** Deterministically format a validated Bible quote (bypasses the LLM). */
export function formatBibleQuote(meta: VerseRecord): string {
  let ref = str(meta.reference_display);
  if (!ref) {
    const book = str(meta.book) || "Genesis";
    const tradBook = str(meta.traditional_book) || TRAD_BOOK_FALLBACK[book.toLowerCase()] || book;
    const chNum = Number(meta.chapter_num ?? 0);
    const gaCh = str(meta.chapter_title_ga) || getGaChapterTitle(chNum);
    const vNum = Number(meta.verse_num ?? 0);
    const gaV = str(meta.ga_label) || getGaLabel(vNum);
    const vRef = str(meta.verse_ref) || `${book} ${chNum}:${vNum}`;
    ref = `${tradBook}, ${gaCh}, ${gaV} (${vRef})`;
  }

  const gaName = str(meta.ga_version_name) || "Ŋmalɛ Krɔŋkrɔŋ Lɛ";
  const gaAbbr = str(meta.ga_version_abbr) || "NEGAB";
  const gaText = str(meta.ga);

  const enName = str(meta.english_version_name) || "King James Version";
  const enAbbr = str(meta.english_version_abbr) || "KJV";
  const enText = str(meta.en);

  let source = str(meta.source_name) || str(meta.filename) || "Bible Archive";
  if (source.includes(".")) source = source.split(".").slice(0, -1).join(".");
  source = source.replace(/_/g, " ").trim();

  return (
    `${ref}\n` +
    `Ga Version: ${gaName} (${gaAbbr}):\n` +
    `${gaText}\n\n` +
    `English Version: ${enName} (${enAbbr}):\n` +
    `${enText}\n\n` +
    `Source: ${source}`
  );
}
